src/ppu/viz/bregman_plot.py

```diff
@@ -0,0 +1,443 @@
+
+
+def draw_classifier(generator, n_samples, names, classifiers, models, rescale = False):
+    dataset = get_dataset(1000, generator, n_samples=n_samples)
+    offset = 1
+    rows = 2
+
+    gridspec_kw={"height_ratios": [1, 1]}
+    cbar_kws = {"use_gridspec": False, "location": "bottom"}
+    plt.rcParams.update({"font.size": 15})
+    figure, axs = plt.subplots(rows, len(models) + offset, gridspec_kw=gridspec_kw, figsize=(27, 6))
+
+    i = 0
+    eps = 1
+    n_ticks = 100
+
+    (X_train, y_train), (X_test, y_test) = dataset  # X is the data point, y is the class label
+
+
+    # the area we draw is a bit larger than the range of data points
+    x_min, x_max = X_train[:, 0].min() - eps, X_train[:, 0].max() + eps
+    y_min, y_max = X_train[:, 1].min() - eps, X_train[:, 1].max() + eps
+
+    # just plot the dataset first
+    cm = plt.cm.RdBu # color map parameter
+    cm_bright = ListedColormap(["#FF0000", "#0000FF"]) # color for data points
+    ax = axs[0][i] # position of subgraph
+    ax.set_title("Data") # subgraph title
+    # Plot the training points
+    ax.scatter(X_train[:, 0], X_train[:, 1], c=y_train, cmap=cm_bright, edgecolors="k")
+    ax.set_xlim(x_min, x_max) # axis range
+    ax.set_ylim(y_min, y_max)
+    ax.set_xticks(())
+    ax.set_yticks(())
+    i += 1
+
+    for name, clf in zip(names, classifiers):
+        ax = axs[0][i]
+        clf.fit(X_train, y_train) # clf is the method in classifiers
+        score = clf.score(X_test, y_test) # score of the result
+        x = np.linspace(x_min, x_max, n_ticks) # for the heatmap we need to creat the grid first, so n_ticks are the density of the grid nodes
+        y = np.linspace(y_min, y_max, n_ticks)
+        xs, ys = np.meshgrid(x, y)
+        X_grid = np.c_[xs.ravel(), ys.ravel()] # All the grid points
+
+        response = _check_boundary_response_method(clf, "auto")(X_grid) # the function returns the probability of points(inputs) belong to each class
+
+        if len(response.shape) != 1:
+            response = response[:, 1] # since there's only 2 classes so 1 can represent the other(the prob sums up to 1)
+
+        display = DecisionBoundaryDisplay(
+            xx0=xs,
+            xx1=ys,
+            response=response.reshape(xs.shape),
+        ) # class to draw DecisionBoundary
+        display.plot(ax=ax, cmap=cm, alpha=0.8) # alpha controls the transparency of the filled contours
+
+        ax.set_xticks(()) # remove the ticks
+        ax.set_yticks(())
+        ax.set_title(name)
+        ax.text(
+            x_max - 0.3,
+            y_min + 0.3,
+            ("%.2f" % score).lstrip("0"),
+            horizontalalignment="right",
+        ) # score as text on the right bottom of the subgraph
+        i += 1
+
+    i = 0
+    ax = axs[1][i]
+    ax.axis("off") # turn off both x-axis and y-axis, including their labels and ticks
+    i += 1
+
+    # iterate over classifiers
+    for name in models:
```

Comment:
This plotting code is reused quite a few times, can we put it in a function to clean it up a bit?

Reply:
For the heat map the process of determining the colour bar is different cross the drawing functions, so I think it may not be able to have a function for all.